arbitrum-docs/for-devs/quickstart-solidity-hardhat.md

```diff
@@ -0,0 +1,368 @@
+   class VendingMachine {
+        // state variables = internal memory of the vending machine
+        cupcakeBalances = {};
+        cupcakeDistributionTimes = {};
+
+        // Vend a cupcake to the caller
+        giveCupcakeTo(userId) {
+            if (this.cupcakeBalances[userId] === undefined) {
+                this.cupcakeBalances[userId] = 0;
+                this.cupcakeDistributionTimes[userId] = 0;
+            }
+
+            // Rule 1: The vending machine will distribute a cupcake to anyone who hasn't recently received one.
+            const fiveSeconds = 5000;
+            const userCanReceiveCupcake = this.cupcakeDistributionTimes[userId] + fiveSeconds <= Date.now();
+            if (userCanReceiveCupcake) {
+                this.cupcakeBalances[userId]++;
+                this.cupcakeDistributionTimes[userId] = Date.now();
+                console.log(`Enjoy your cupcake, ${userId}!`);
+                return true;
+            } else {
+                console.error("HTTP 429: Too Many Cupcakes (you must wait at least 5 seconds between cupcakes)");
+                return false;
+            }
+        }
+
+        getCupcakeBalanceFor(userId) {
+            return this.cupcakeBalances[userId];
+        }
+    }
+```
+
+The `VendingMachine` class uses *state variables* and *functions* to implement *predefined rules*. This implementation is useful because it automates cupcake distribution, but it's "dumb" because it's hosted by a centralized server that we (Offchain Labs) control. So it's technically possible for someone at Offchain to give their friends extra cupcakes by modifying the above contract's code.
+
+This is a problem because it means that you have to trust us, and we have to trust our hosting provider. How do we keep our cupcakes secure without having to trust anyone?
+
+DApps solve this problem by using Ethereum smart contracts to build <a data-quicklook-from='trustless'>*trustless*</a> end-user experiences that nobody in particular controls[^4]. This gives us a way to follow the second rule: *"The vending machine's rules can't be changed by anyone."*[^5]
+
+Let's convert our dumb contract into a smart contract.
+
+
+### Configure your project directory
+
+Create a `decentralized-cupcakes` directory for your project and install [hardhat](https://hardhat.org/hardhat-runner/docs/getting-started#overview)[^6] using VS Code's integrated terminal:
+
+```
+mkdir decentralized-cupcakes
+cd decentralized-cupcakes
+yarn init -y
+yarn add hardhat @nomicfoundation/hardhat-toolbox
+```
+
+This installs two packages: `hardhat` lets us test and deploy our smart contracts, and `hardhat-toolbox` is a bundle of popular Hardhat plugins that we'll use later.
+
+Next, run `npx hardhat` to configure Hardhat. Select `Create a JavaScript project` when prompted. Make sure you specify your `decentralized-cupcakes` directory as the project root when asked.
```

Comment:
I would stick with one package manager: either yarn or npm for the whole tutorial. Seeing that you only mention yarn when installing the package, it might be easier to just replace that with npm.
The alternative is to replace "npx" with "yarn" here and in the next occurrences.

Reply:
For now, the most valuable form of feedback that you can provide is feedback related to the broad strokes of the procedure, and the code - I may end up implementing toggles so users can switch between consistent `npm` vs `yarn` instructions, but that's TBD, and in the meantime I'm recording my procedure :). 

I'll flip this from `Draft` to `Open` when I'm ready for micro feedback. Thank you!